Add app code for cell voltage readings, remove cell temp and die temp temporarily

pasitlaothamatas opened the pull request:
### Summary
<!-- Quick summary of changes, optional -->
Wanted to get new app code integrated so i can test io cell temp and io die temp more easily

- Add app code for cell voltage measurement
- Remove cell temp and die temp measurements
- Remove can messages associated with removed cell temp and die temp measurements
- Binary status is pretty unnecessary imo, not used currently and should not be used in the future. Removed

- Create new directory for LTC related IO files
- Added documentation and small fixes for Io_CellVoltages

### Changelist 
<!-- Give a list of the changes covered in this PR. This will help both you and the reviewer keep this PR within scope. -->

### Testing Done
<!-- Outline the testing that was done to demonstrate the changes are solid. This could be unit tests, integration tests, testing on the car, etc. Include relevant code snippets, screenshots, etc as needed. -->
- tested that cell voltages for a single segment can be read back
- tested that dcc bits can still be written to

### Resolved Issues
<!-- Link any issues that this PR resolved like so: `Resolves #1, #2, and #5` (Note: Using this format, Github will automatically close the issue(s) when this PR is merged in). -->

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [ ] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [ ] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

timbitss commented on boards/BMS/Inc/App/App_Accumulator.h:
```diff
@@ -1,185 +1,48 @@
+    bool (*configure_cell_monitors)(void),
+    bool (*start_voltage_conv)(void),
+    bool (*read_cell_voltages)(void),
+    void (*get_min_cell_location)(uint8_t *, uint8_t *),
+    void (*get_max_cell_location)(uint8_t *, uint8_t *),
```
What are your thoughts on including the names of the function pointer arguments? It's not clear what the two uint8_t ptrs are supposed to represent

pasitlaothamatas replied:
i think we can refer to the io function that we currently pass in right now

timbitss commented on boards/BMS/Inc/App/App_SetPeriodicCanSignals.h:
```diff
@@ -5,12 +5,3 @@
     struct BmsCanTxInterface *can_tx,
     struct Imd *              imd);
-
-void App_SetPeriodicSignals_AccumulatorInRangeChecks(
-    struct BmsCanTxInterface *can_tx,
-    const struct Accumulator *accumulator,
-    struct ErrorTable *       error_table);
-
-void App_SetPeriodicSignals_CellMonitorsInRangeChecks(
-    struct BmsCanTxInterface * can_tx,
-    const struct CellMonitors *cell_monitors);
```
Why was this deleted?

pasitlaothamatas replied:
Bc we shouldnt use in range checks, i think theyre a little extra

pasitlaothamatas replied:
I also want to wrap these using functions from App_Accumulator, don't think we need an entirely new file containing functions to broadcast CAN messages

timbitss reviewed (commented):
thoughts??

timbitss commented on boards/BMS/Inc/Io/Io_LTC6813/Io_SharedLTC6813.h:
```diff
@@ -45,21 +45,21 @@ enum RegGroupByteFormat
  * @param tx_data The command or register group to pack PEC15 bytes
  * @param size The size of the command or register group to transmit
  */
```
rename to `num_bytes` and update comment

maybe mention that it's the number of bytes to transmit without the PEC15 since from first glance I would think it's the size of the tx_data buffer

pasitlaothamatas replied:
ah yeah nice catch, looks like some comments i forgot to update

timbitss commented on boards/BMS/Inc/Io/Io_LTC6813/Io_SharedLTC6813.h:
```diff
@@ -68,22 +68,22 @@ void Io_NewLTC6813_InitSpiHandle(SPI_HandleTypeDef *spi_handle);
  * @return The PEC15 code generated from the contents of data_buffer
  */
-uint16_t Io_NewLTC6813_CalculateRegGroupPec15(uint8_t *data_buffer);
+uint16_t Io_SharedLTC6813_CalculateRegGroupPec15(uint8_t *data_buffer);
```
should make pointer const since we're not modifying the data buffer

pasitlaothamatas replied:
im assuming you meant pointer to a `const uint8_t` here, added

timbitss commented on boards/BMS/Inc/Io/Io_LTC6813/Io_SharedLTC6813.h:
```diff
@@ -45,21 +45,21 @@ enum RegGroupByteFormat
  * @param spi_handle The given SPI handle for the LTC6813 daisy chain.
  */
-void Io_NewLTC6813_InitSpiHandle(SPI_HandleTypeDef *spi_handle);
+void Io_SharedLTC6813_InitSpiHandle(SPI_HandleTypeDef *spi_handle);
```
I think we should put the init function at the top as an indication that it should called on startup

pasitlaothamatas replied:
hm ya agree

timbitss commented on boards/BMS/Src/App/App_Accumulator.c:
```diff
@@ -1,160 +1,89 @@
 struct Accumulator *App_Accumulator_Create(
-    ExitCode (*configure_cell_monitors)(void),
-    ExitCode (*read_cell_voltages)(void),
+    bool (*configure_monitoring_chip)(void),
```
parameter name `configure_monitoring_chip` should match declaration in header file, could be more explicit that it's the LTC chip in particular

pasitlaothamatas replied:
ah yeah need to change this in the struct as well. i feel like running misra checks will help us a lot with detecting these. we should look into maybe doing this at some point

pasitlaothamatas replied:
i would prefer to keep sensor information abstracted from these fxn ptrs though, since the app layer shouldnt really have sensor specific information. Rest of the app code for the vehicle follows this convention so it think its also nice to keep things consistent

Taken from the FSM for example, specific sensor information is abstracted here.
![image](https://user-images.githubusercontent.com/32230980/158759519-e927a3f9-933f-4b0c-8687-c1372936fa3b.png)

timbitss commented on boards/BMS/Src/Io/Io_LTC6813/Io_SharedLTC6813.c:
```diff
@@ -182,7 +190,7 @@ static void Io_PrepareCfgRegBytes(
-    Io_NewCellVoltages_GetMinCellLocation(&min_cell_segment, &min_cell_index);
+    Io_CellVoltages_GetMinCellLocation(&min_cell_segment, &min_cell_index);
 
     // Write to the last segment on the daisy chain first
```
Does the last segment on the daisy chain correspond to segment 0?

pasitlaothamatas replied:
yeah i think i see where your confusion is coming from

this is rlly stupid i think i have a better way to do this. stay tuned
![image](https://user-images.githubusercontent.com/32230980/158760377-07bbb72e-e3dc-49ab-ba99-b36d1732c93b.png)

pasitlaothamatas replied:
implementation should be much nicer now

timbitss commented on boards/BMS/Src/Io/Io_LTC6813/Io_SharedLTC6813.c:
```diff
@@ -233,48 +242,48 @@ void Io_NewLTC6813_InitSpiHandle(SPI_HandleTypeDef *spi_handle)
-void Io_NewLTC6813_PackPec15(uint8_t *tx_data, uint8_t num_bytes)
+void Io_SharedLTC6813_PackPec15(uint8_t *tx_data, uint8_t num_bytes)
 {
     const uint16_t cfg_reg_a_pec15 =
```
I think naming the var `cfg_reg_a_pec15` is a bit unclear, simply naming it `pec15` would make more sense

pasitlaothamatas replied:
ah yea, good catch. looks like some leftover var names i can fix this.

pasitlaothamatas commented on boards/BMS/Src/App/states/App_AllStates.c:
```diff
@@ -24,11 +25,12 @@ void App_AllStatesRunOnTick100Hz(struct StateMachine *const state_machine)
     struct ErrorTable *       error_table = App_BmsWorld_GetErrorTable(world);
 
     App_SetPeriodicCanSignals_Imd(can_tx, imd);
+    App_Accumulator_RunOnTick100Hz(accumulator, can_tx);
```
whoops need to move this

timbitss commented on boards/BMS/Src/Io/Io_LTC6813/Io_SharedLTC6813.c:
```diff
@@ -233,48 +242,48 @@ void Io_NewLTC6813_InitSpiHandle(SPI_HandleTypeDef *spi_handle)
 }
 
-void Io_NewLTC6813_PrepareCmd(uint16_t *tx_cmd, uint16_t cmd)
+void Io_SharedLTC6813_PrepareCmd(uint16_t *tx_cmd, uint16_t cmd)
```
What do you think of renaming it to `Io_SharedLTC6813_PrepareCmdWithPEC15`

pasitlaothamatas replied:
this isnt just preparing with pec15 though. its changing the cmd endianness and then doing the pec15 check.

timbitss commented on boards/BMS/Src/Io/Io_LTC6813/Io_CellVoltages.c:
```diff
@@ -176,7 +190,7 @@ bool Io_NewCellVoltages_ReadVoltages(void)
+            Io_SharedLTC6813_PrepareCmd(tx_cmd, cv_read_cmds[curr_reg_group]);
 
             // Transmit the command and receive data stored in register group
             if (Io_SharedSpi_TransmitAndReceive(
```
Could we possibly wrap `Io_SharedSpi_TransmitAndReceive()` into `Io_SharedLTC6813`? e.g. `Io_SharedLTC6813_SendReadCmd()`. That way, we won't need to expose the SPI handle outside of `Io_SharedLTC6813.c`

pasitlaothamatas replied:
Yeah im thinking about this too. We can actually get rid of shared spi entirely and implement SPI commands within the IO_SharedLTC file since the BMS is the only board that uses it. Old PDM needed the SPI driver, but new PDM no longer needs a SPI interface. Lets leave this for another PR though. I think its out of scope rn 🅿️🅿️🅿️

timbitss replied:
LOL

timbitss reviewed (commented):
Looks good. See comments below

pasitlaothamatas commented on boards/shared/Inc/App/App_SharedSetPeriodicCanSignals.h:
```diff
@@ -35,21 +35,3 @@
         return status;                                                         \
     }
-
-#define STATIC_DEFINE_APP_SET_PERIODIC_CAN_SIGNALS_BINARY_STATUS(            \
-    CAN_TX_INTERFACE)                                                        \
-    static void App_SetPeriodicCanSignals_SharedBinaryStatus(                \
-        struct CAN_TX_INTERFACE *  can_tx,                                   \
-        const struct BinaryStatus *binary_status,                            \
-        void (*const can_signal_setter)(struct CAN_TX_INTERFACE *, uint8_t), \
-        uint8_t on_choice, uint8_t off_choice)                               \
-    {                                                                        \
-        if (App_SharedBinaryStatus_IsActive(binary_status))                  \
-        {                                                                    \
-            can_signal_setter(can_tx, on_choice);                            \
-        }                                                                    \
-        else                                                                 \
-        {                                                                    \
-            can_signal_setter(can_tx, off_choice);                           \
-        }                                                                    \
-    }
```
Sure i can do that. Theyre currently not used in the BMS project, and i dont plan on integrating them in the future

pasitlaothamatas commented on scripts/codegen/CAN/App_CanMsgs.dbc:
```diff
@@ -80,55 +80,6 @@ SG_ AIR_NEGATIVE : 1|1@1+ (1,0) [0|1] "" FSM,DCM,PDM
-
-BO_ 117 BMS_ACCUMULATOR_SEGMENT_0: 4 BMS
-SG_ SEGMENT_0_VOLTAGE : 0|32@1+ (1,0) [48.0|67.2] "V" DEBUG
-
-BO_ 118 BMS_ACCUMULATOR_SEGMENT_1: 4 BMS
-SG_ SEGMENT_1_VOLTAGE : 0|32@1+ (1,0) [48.0|67.2] "V" DEBUG
-
-BO_ 119 BMS_ACCUMULATOR_SEGMENT_2: 4 BMS
-SG_ SEGMENT_2_VOLTAGE : 0|32@1+ (1,0) [48.0|67.2] "V" DEBUG
-
-BO_ 120 BMS_ACCUMULATOR_SEGMENT_3: 4 BMS
-SG_ SEGMENT_3_VOLTAGE : 0|32@1+ (1,0) [48.0|67.2] "V" DEBUG
-
-BO_ 121 BMS_ACCUMULATOR_SEGMENT_4: 4 BMS
-SG_ SEGMENT_4_VOLTAGE : 0|32@1+ (1,0) [48.0|67.2] "V" DEBUG
-
-BO_ 122 BMS_ACCUMULATOR_SEGMENT_5: 4 BMS
-SG_ SEGMENT_5_VOLTAGE : 0|32@1+ (1,0) [48.0|67.2] "V" DEBUG
-
-BO_ 123 BMS_CELL_MONITOR_0: 4 BMS
-SG_ CELL_MONITOR_0_DIE_TEMPERATURE : 0|32@1+ (1,0) [0.0|120.0] "degC" DEBUG
-
-BO_ 124 BMS_CELL_MONITOR_1: 4 BMS
-SG_ CELL_MONITOR_1_DIE_TEMPERATURE : 0|32@1+ (1,0) [0.0|120.0] "degC" DEBUG
-
-BO_ 125 BMS_CELL_MONITOR_2: 4 BMS
-SG_ CELL_MONITOR_2_DIE_TEMPERATURE : 0|32@1+ (1,0) [0.0|120.0] "degC" DEBUG
-
-BO_ 126 BMS_CELL_MONITOR_3: 4 BMS
-SG_ CELL_MONITOR_3_DIE_TEMPERATURE : 0|32@1+ (1,0) [0.0|120.0] "degC" DEBUG
-
-BO_ 127 BMS_CELL_MONITOR_4: 4 BMS
-SG_ CELL_MONITOR_4_DIE_TEMPERATURE : 0|32@1+ (1,0) [0.0|120.0] "degC" DEBUG
-
-BO_ 128 BMS_CELL_MONITOR_5: 4 BMS
-SG_ CELL_MONITOR_5_DIE_TEMPERATURE : 0|32@1+ (1,0) [0.0|120.0] "degC" DEBUG
-
-BO_ 129 BMS_MAX_CELL_MONITOR: 4 BMS
-SG_ MAX_CELL_MONITOR_DIE_TEMPERATURE : 0|32@1+ (1,0) [0.0|120.0] "degC" DEBUG
-
```
die temp and cell temp code, as well as the app level wrapper needs to be updated in future PR. They are removed for the current PR so these can messages should also be removed. I can add this to the PR description

pasitlaothamatas commented on boards/BMS/Src/Io/Io_LTC6813/Io_CellVoltages.c:
```diff
@@ -59,10 +62,21 @@ static uint16_t        cell_voltages[NUM_OF_ACCUMULATOR_SEGMENTS]
+ * @param rx_buffer The buffer containing data read from the LTC6813 chips
+ * @return True if the data is read back successfully. Else, false
+ */
+static bool Io_ParseCellVoltageFromAllSegments(
```
sure
```suggestion
static bool Io_ParseCellVoltagesFromAllSegments(
```

pasitlaothamatas commented:
@timbitss Went over the comments, should all be addressed. 

also did the following:
- Renamed the Io files
- removed some unnecessary fxns in the `Io_LTC6813Shared` files
- Cleaned up `#defines` and `enums` in `Io_LTC6813` dir
- Clean up some unnecessary CAN msgs in the dbc
- Reversed the endianness of the commands used to talk to the LTC chip. Better to store them in their reversed endianness rather than storing them similarly to how they are presented in the datasheet as we can avoid swapping the endianness in runtime imo

timbitss reviewed (approved):
Good work :)

pasitlaothamatas commented:
> Good work :)



> Good work :)

lfg

pasitlaothamatas commented:
> Good work :)



> Good work :)

lfg
